Additional validation from the wat parser

ColinEberhardt opened the pull request:
while running `wat2wasm` over various test cases I've spotted some rules that are not being applied. Currently there is validation logic within the interpreter, but not the wat-parser. As these rules seems to be enforced by `wat2wasm` I thought it was best to enforce them in the parser.

What do you think?

I've got a few other tests to fix as well, but thought it was worth getting some feedback.

maurobringolf commented:
Awesome, this looks great to me. Are these specific to wat or could these be placed in one of the other packages to be reused?

ColinEberhardt commented:
> Awesome, this looks great to me. Are these specific to wat or could these be placed in one of the other packages to be reused?

Good question, however, the structure of wasm, with clearly defined sections, might require a different approach to validation.

> Related: I have been reading a bit on validation of types and if I understand correctly it should be possible to implement a pass that verifies that all instruction stack argument and return value type match up. Do we have anything for this setup?

That's a very good point, we certainly have some wat files in the interpreter tests that fail this check ;-)

I'm planning to add additional validation as I go along - the main thing to debate initially is whether this belongs in wat-parser.

xtuc commented:
> Related: I have been reading a bit on validation of types and if I understand correctly it should be possible to implement a pass that verifies that all instruction stack argument and return value type match up. Do we have anything for this setup?

Yes, I started it here https://github.com/xtuc/webassemblyjs/blob/master/packages/webassemblyjs/src/compiler/validation/type-inference.js.

xtuc commented on packages/wast-parser/src/validator.js:
```diff
@@ -0,0 +1,47 @@
+  });
+
+  if (startFuncs > 1) {
+    throw new Error("wat modules can only contain one Start function");
```
For my own debugging purposes, I started using the `babel-code-frame` package which allows us to print the location in the code along with the error. In the future we could re-implement it with our own printer or with more information in the parser.

ColinEberhardt commented:
Thanks for the feedback @xtuc & @maurobringolf what's the best route forward for this PR?

> Looks great! I'm wondering if it would make sense to have a new package @webassemblyjs/validations that provide these validations?

Should I migrate these validations, and those from the interpreter, before merging?

xtuc commented:
> Should I migrate these validations, and those from the interpreter, before merging?

We an do it after your merge, I can also create the package if you want.

ColinEberhardt commented:
>> Should I migrate these validations, and those from the interpreter, before merging?

> We an do it after your merge, I can also create the package if you want.

OK, cool - I'll fix up the tests and merge. Thanks

ColinEberhardt commented:
@xtuc I've fixed up the tests, but this uncovered a few more issues which i have resolved. You might want to review again before I merge.

The changes are:
- AST traversal additional recursion to create a proper path for module import nested statements
- with global shorthand imports the parser previously moved the import outside of the global within the AST. I can't see any way the relationship between the two was preserved? I've modified this behaviour so that the import is nested within the global in the `init` property. 

I also think the `init` property of a global should not be an array, as per the rules discussed here: https://github.com/WebAssembly/spec/issues/694 - but I've not made that change yet.

xtuc commented on packages/wast-parser/src/index.js:
```diff
@@ -1,13 +1,14 @@
   // We pass the source here to show code frames
   const ast = parser.parse(tokens, source);
-
+  validate(ast);
```
Actually, I don't know what's the best place for this check. I used to run the validations here but it annoyed me when I wanted to just test the parser. I moved it right in the module "compilation".

What do you think? It's not a parse error but a validation one.

ColinEberhardt replied:
Yeah, that’s a good point, having validation in multiple places will confused things. I initially added it here to mirror the functionality of wat2wasm,but that doesn’t mean it’s right for this project. I’ll move it :-)

xtuc replied:
ok thanks

ColinEberhardt commented on packages/webassemblyjs/test/index.js:
```diff
@@ -15,7 +15,7 @@ function toArrayBuffer(buf) {
   describe("wasm", () => {
     const testSuites = glob.sync(
-      "packages/WebAssembly/test/fixtures/**/module.wasm"
+      "packages/webassemblyjs/test/fixtures/**/module.wasm"
```
@xtuc I was wondering why the total number of unit tests had dropped recently ;-)

xtuc replied:
omg, that's a good catch!
